### ISSUE TITLE
High CPU and low throughput with SSL torrents 

### ISSUE THREAD

**Issue by rageshkrishna** [arvidn/libtorrent#7433]
I wrote a bare minimum libtorrent peer and measured the time taken to transfer a single 1600MB file between two peers on Digital Ocean.

For a regular torrent, it took 7s and used the entire 225MBps throughput between these machines that I measured with iperf3. The same file as an SSL torrent took 20s with significantly higher CPU usage. Roughly 150% more. 

I'll be happy to provide more information if it helps.

**Comment by arvidn** [comment]
I believe that the problem is that all SSL crypto is performed in the network thread, which is only a single thread.
I made an attempt back in libtorrent-1.1 to offload just the `async_read()` and `async_write()` calls to a thread pool, but that ended up causing some races and heap corruption.

**Comment by rageshkrishna** [comment]
Thanks for the input, @arvidn. This seems like a very intricate issue. Unfortunately I'm not familiar with that part of the code yet. I'll start educating myself and see if I can take a stab at it.

**Comment by ValeZAA** [comment]
>The same file as an SSL torrent took 20s with significantly higher CPU usage. Roughly 150% more.

And so what? Higher CPU usage does not mean slower 20 seconds speeds.

Or you mean IT SHOULD take even more CPU cycles if it needs 7 seconds back? What is the ACTUAL bug here? That HW acceleration of SSL enxryption is broken?

**Comment by rageshkrishna** [comment]
Yes, it does look like a faster CPU improves the download speed. I tested on `n2` machines on GCP and I think it took around 13s there to transfer the same content (sorry I don't remember the exact details of the machine config at this point).

I'm not sure if I'd call it a bug, it's more of a "performance" issue. The same machine can transfer the same file over HTTPS much, much faster than libtorrent + SSL so I'd say that that SSL by itself isn't the problem here.

I fully agree that bittorrent itself might have more overhead than HTTPS, but going from 7s to 20s just by turning on SSL functionality seems excessive. It's clear that something in the SSL code paths is slowing it down, but as I said earlier, I'm not familiar with that code to comment on the internals.

**Comment by XinfitX** [comment]
In a quick test after my ISP migrated to XGS-PON I hired 10Gbps, I could notice that at high download / upload rates I noticed that the core goes to 100%, in a test in my network I could notice a high clock CPU - 4 cores it has much better performance compared to CPU with many cores, 16 cores and lower clock.
@arvidn is there any way to make libtorrent more efficient for multi-core CPU? 10Gbps becoming reality very fast with xgs-pon

**Comment by Rootax** [comment]
> In a quick test after my ISP migrated to XGS-PON I hired 10Gbps, I could notice that at high download / upload rates I noticed that the core goes to 100%, in a test in my network I could notice a high clock CPU - 4 cores it has much better performance compared to CPU with many cores, 16 cores and lower clock. @arvidn is there any way to make libtorrent more efficient for multi-core CPU? 10Gbps becoming reality very fast with xgs-pon

You would need to split / thread the network thread afaik. It would be a huge work. And you have to have a robust storage i/o too (in my case, I stayed with 1.x libt, since 2.0 had horrible disk performances for me. Maybe it's better now...)
